src/VecSim/algorithms/hnsw/hnsw_tiered.h

```diff
@@ -870,47 +871,51 @@ TieredHNSWIndex<DataType, DistType>::TieredHNSW_BatchIterator::~TieredHNSW_Batch
     }
 
     this->allocator->free_allocation(this->queryParams);
-
-    VecSimQueryResult_Free(this->flat_results);
-    VecSimQueryResult_Free(this->hnsw_results);
 }
 
 /******************** Implementation **************/
 
 template <typename DataType, typename DistType>
-VecSimQueryResult_List
+VecSimQueryResult_List *
 TieredHNSWIndex<DataType, DistType>::TieredHNSW_BatchIterator::getNextResults(
     size_t n_res, VecSimQueryResult_Order order) {
 
     const bool isMulti = this->index->backendIndex->isMultiValue();
+    auto hnsw_code = VecSim_QueryResult_OK;
 
     if (this->hnsw_iterator == UNINITIALIZED) {
         // First call to getNextResults. The call to the BF iterator will include calculating all
         // the distances and access the BF index. We take the lock on this call.
         this->index->flatIndexGuard.lock_shared();
-        this->flat_results = this->flat_iterator->getNextResults(n_res, BY_SCORE_THEN_ID);
+        auto cur_flat_results = this->flat_iterator->getNextResults(n_res, BY_SCORE_THEN_ID);
         this->index->flatIndexGuard.unlock_shared();
         // This is also the only time `getNextResults` on the BF iterator can fail.
-        if (VecSim_OK != flat_results.code) {
-            return flat_results;
+        if (VecSim_OK != cur_flat_results->code) {
+            return cur_flat_results;
         }
```

Comment:
should this->flat_results be updated?

Reply:
No.
1. We return them all. `this->flat_results` holds results that are "cached" for the next batch
2. We got a timeout. Most likely the iterator is going to be freed soon